### ISSUE TITLE
Specify LATENT at the CREATE GENERATOR, not CREATE POPULATION, level

### ISSUE THREAD

**Issue by fsaad** [probcomp/bayeslite#435]
Exposing a latent variable at the population level is the wrong level of abstraction. A statistical population is a set of observable variables with annotated statistical types. This characterization is model independent.

The notion of an "exposed latent variable", however, is model-specific. The latents that can be exposed are a function of the cgpms that override the default model in the statement `CREATE GENERATOR ... (MODEL x, y USING foo)`, where `y` is a latent available from `foo`. Pre-specifying the latents at the population level breaks this independence.

Take for example, the following MML code

```
**CREATE POPULATION satellites FOR satellites_ucs (
    Name                            IGNORE,
    Country_of_Operator             CATEGORICAL,
    Operator_Owner                  CATEGORICAL,
    Users                           CATEGORICAL,
    Purpose                         CATEGORICAL,
    Class_of_Orbit                  CATEGORICAL,
    Type_of_Orbit                   CATEGORICAL,
    Perigee_km                      NUMERICAL,
    Apogee_km                       NUMERICAL,
    Eccentricity                    NUMERICAL,
    Period_minutes                  NUMERICAL,
    Launch_Mass_kg                  NUMERICAL,
    Dry_Mass_kg                     NUMERICAL,
    Power_watts                     NUMERICAL,
    Date_of_Launch                  NUMERICAL,
    Anticipated_Lifetime            NUMERICAL,
    Contractor                      CATEGORICAL,
    Country_of_Contractor           CATEGORICAL,
    Launch_Site                     CATEGORICAL,
    Launch_Vehicle                  CATEGORICAL,
    Source_Used_for_Orbital_Data    CATEGORICAL,
    longitude_radians_of_geo        CYCLIC,
    Inclination_radians             CYCLIC,

    LATENT kepler_cluster           CATEGORICAL,
    LATENT kepler_noise             NUMERICAL
);

CREATE GENERATOR satellites_default FOR satellites USING cgpm (
    SUBSAMPLE 100

```

The `satellites_default` cgpm is going to have no notion of the latent variables (it cannot model them), a very confusing situation. However, now we have a second generator

```
CREATE GENERATOR satellites_custom FOR satellites USING cgpm (
    apogee_km    LOGNORMAL,
    perigee_km   LOGNORMAL,

    MODEL kepler_cluster, kepler_noise, period_minutes
        GIVEN apogee_km, perigee_km
        USING venturescript (sp = kepler),

    MODEL anticipated_lifetime
        GIVEN date_of_launch, power_watts, apogee_km, perigee_km,
            dry_mass_kg, class_of_orbit
        USING linear_regression,

    MODEL class_of_orbit
        GIVEN apogee_km, perigee_km, period_minutes
        USING random_forest (k = 4),

    SUBSAMPLE 100
);
```

which is modeling the exposed latents using the venturescript CGPM. The solution is to expose the latents in the MML code for `satellites_custom`, i.e.

```
CREATE GENERATOR satellites_custom FOR satellites USING cgpm (
    apogee_km    LOGNORMAL,
    perigee_km   LOGNORMAL,

    LATENT kepler_cluster CATEGORICAL;
    LATENT kepler_noise NUMERICAL;

    MODEL kepler_cluster, kepler_noise, period_minutes
        GIVEN apogee_km, perigee_km
        USING venturescript (sp = kepler),

    MODEL anticipated_lifetime
        GIVEN date_of_launch, power_watts, apogee_km, perigee_km,
            dry_mass_kg, class_of_orbit
        USING linear_regression,

    MODEL class_of_orbit
        GIVEN apogee_km, perigee_km, period_minutes
        USING random_forest (k = 4),

    SUBSAMPLE 100
);
```

Design considerations:

- Any BQL query against a population which aggregates over generator can only include variable names that are in the population and all generators agree on.
- Queries on latent variables must specify the generator that expose those latents.

This change is necessary for (a) release of the notebook, and (b) code required for implementation of crosscat, k-nn, pca, and k-means CGPMs (all with different latent variables) against a common base population next week.

**Comment by riastradh-probcomp** [comment]
First: I don't think this should block any 0.2.0 release or public demo -- we are already snowballing what should have been out last week.

That said, there were a couple of reasons motivating the inclusion of latents in a population:

1. It was not a priori clear to me that latents would never be shared between generators.  E.g., maybe the Kepler noise can be modelled two different ways.

2. Making some variables per-generator requires much more substantial changes to the way BQL is implemented.  There was absolutely no way that was going to happen last week; it could easily delay 0.2 by a month now, or at least week if I ignore everything else.

**Comment by fsaad** [comment]
> we are already snowballing 

Can you please outline what you feel we are snowballing?

> what should have been out last week.

I suggest not worry or become anxious about this point. There is no "external" pressure to rush up the release. If you feel there is internal pressure, then we should talk about it.

> Making some variables per-generator requires much more substantial changes to the way BQL is implemented.

The quick solution is to support the change cosmetically until we more substantially define populations and generators:

1. Extend the grammar for cgpm_schema to include LATENT variable.
2. In `create_generator`, update the table `bayesdb_variable` as is currently done in `create_population`.

I can do the above changes, probably should take 2/3 hours.

**Comment by riastradh-probcomp** [comment]
On reflection, I see now that it probably won't take as much work as I first estimated to get per-generator latents -- specifically, I figured it would be necessary to do https://github.com/probcomp/bayeslite/issues/38 first, but the only places latents will be usable is places where we already have to do a minimal level of lexical scope tracking already: DEPENDENCE PROBABILITY OF x WITH y, PROBABILITY OF x = v, &c.

So we should add the column

    generator_id    INTEGER REFERENCES bayesdb_generator(id) ON DELETE CASCADE

to the bayesdb_variable table; define an operation, say `bayesdb_add_latent`, in bayeslite.core, to add a latent variable with a given statistical type; and teach cgpm's `create_generator` to call it as you suggested.

We'll have to decide whether to require an explicit `MODELLED BY` clause, or just pick the unique generator that works, in any BQL query that uses latent variables -- this will have to be detected in compiler.py.  I'm tempted to say we should require `MODELLED BY` for the moment.
